astacus/coordinator/plugins/clickhouse/object_storage.py

```diff
@@ -91,12 +91,24 @@ def _copy_items_between(
         target_storage: BaseTransfer[Any],
         stats: StatsClient | None,
     ) -> None:
+        total_keys = len(keys)
         match self.copy_config:
             case DirectCopyConfig():
-                logger.info("Copying the keys using the cloud APIs")
-                target_storage.copy_files_from(source=source_storage, keys=keys)
+                logger.info("Copying %i keys using the cloud APIs", total_keys)
+                target_storage.copy_files_from(
+                    source=source_storage,
+                    keys=keys,
+                    progress_fn=lambda completed_files, total_files: _handle_copy_progress_fn(
```

Comment:
discussed in more detail, but basically I want some changes to the code structure:
* rename the handler to `emit_.._metric` to declare intent
* hide the details of direct copy into a function similar to _copy_via_local to keep the match statement clear and forgettable
* maybe use functools.partial instead of lambda to bind method and stats

Reply:
See https://github.com/Aiven-Open/astacus/pull/229#discussion_r1734481164